main/src/ca/uwaterloo/flix/language/phase/unification/set/SetUnification.scala

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2024 Jonathan Lindegaard Starup
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.unification.set
+
+import SetFormula.*
+
+object SetUnification {
+
+  /** A class to track progress during set unification. */
+  class Progress private(var somethingHappened: Boolean = false) {
+    /** Mark that progress has been made. */
+    def progress(): Unit = somethingHappened = true
+
+    /** Returns `true` if there has been progress. */
+    def hasProgressed: Boolean = somethingHappened
+  }
+
+  object Progress {
+    /** Make a new [[Progress]] object with progress set to `false`. */
+    def apply(): Progress = new Progress()
+  }
+
+  /** A collection of rules that can be used to solve individual [[Equation]]. */
+  private object Rules {
+
+    /**
+      * Solves equations that trivially hold (like `univ ~ univ`) and marks trivially unsolvable
+      * equations (like `univ ~ empty`). Always returns an empty substitution.
+      *
+      * Otherwise returns `eq` directly.
+      *
+      *   - `univ ~ univ` becomes `{}`
+      *   - `univ ~ empty` becomes `{univ !~ empty}`
+      *   - `x1 ~ x1` becomes `{}`
+      *   - `x2 ~ univ` becomes `{x2 ~ univ}`
+      */
+    def trivial(eq: Equation)(implicit p: Progress): (List[Equation], SetSubstitution) = {
+      val Equation(t1, t2, _, _) = eq
+
+      @inline
```

Comment:
Note that: Inlining these probably does little for performance. However, it does increase the size of `trivial` so that it cannot be inlined.

Reply:
> [..] so that it cannot be inlined.

I mean who knows the size threshold? might also not be inlined anyway and the straightline, inlined program is faster?

Reply:
There is a book on Jvm Internals that you can read :) I read it, but cannot remember the name. Just saying, dont go by folklore. At least you should bench it.

Reply:
Maybe its this one: https://www.amazon.com/JVM-Performance-Engineering-OpenJDK-Developers-ebook/dp/B08F5J6V4T but I am not sure

Reply:
right, my intention was not to performance game, my intention was just to reduce clutter for the reader without changing its actual behaviour/performance. I just wanted "syntactic functions" if that makes sense. That's why I felt that `@inline` was the neutral choice. But I can remove it